### ISSUE TITLE
MAXDOP is set per Processor not Task

### ISSUE THREAD

**Issue by dinesh1440** [MicrosoftDocs/sql-docs#4207]
Hi,
Through my experience I have seen MAXDOP is set per processor.  A SQL Server with 8 Schedulers can have 32 child tasks and one parent tasks, which will mean 33 tasks across 8 schedulers With Maxdop set to 8. 
So I feel that the statement, MAXDop is set per task is incorrect. Please let me know if the understanding is correct

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5e3528c6-06d5-aaef-58ac-a4426ca2a433
* Version Independent ID: 6264cc46-28bb-c4d6-067d-67c9c417f4b1
* Content: [Configure the max degree of parallelism Server Configuration Option - SQL Server](https://docs.microsoft.com/en-us/sql/database-engine/configure-windows/configure-the-max-degree-of-parallelism-server-configuration-option?view=sql-server-ver15#feedback)
* Content Source: [docs/database-engine/configure-windows/configure-the-max-degree-of-parallelism-server-configuration-option.md](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/database-engine/configure-windows/configure-the-max-degree-of-parallelism-server-configuration-option.md)
* Product: **sql**
* Technology: **configuration**
* GitHub Login: @MikeRayMSFT
* Microsoft Alias: **mikeray**

**Comment by WilliamAntonRohm** [comment]
@dinesh1440 -- thank you for your feedback.

@MikeRayMSFT -- please look into this issue #4207.

**Comment by pmasl** [comment]
Take the example of a query (request) that only has a merge join between 2 indexes, and both the join and the index reads are all executed in parallel with a maxdop 32 limit on an 8 cpu box. Each index operation + the join will have a coordinating task and a max of 32 other tasks associated (hence the maxdop limit being per task, not per request). At any given time, this request has at least 99 tasks executing. Each task is 1:1 with a worker thread, which during its running state, is assigned to a scheduler (CPU). Because we only have 8 schedulers, any running task will yield it's scheduler time (quantum) so that a runnable task can then be running (assuming there are no waits, in which case the task would be suspended).
This means that during a parallel query execution, a single request can spawn multiple tasks, and each task is assigned to a worker, up to the MAXDOP limit. 

See the https://docs.microsoft.com/en-us/sql/relational-databases/query-processing-architecture-guide and https://docs.microsoft.com/en-us/sql/relational-databases/thread-and-task-architecture-guide?view=sql-server-ver15#sql-server-task-scheduling for details.

**Comment by dinesh1440** [comment]
Thanks Pedro.

I thought It was per Processor. Anyway thanks for the clarification.

But I would like to understand, what is the need of having more tasks than
number of processors. Considering at a time only 8 tasks can be processed
in a 8 Logical CPUs box, how the tasks waiting in the queue on the same
processor for the same query help the parelleism!? Spanning 32 tasks means
4 tasks per Processor. So 1 will be running while 3 will be waiting. And
this adds overhead in context switching, which could be avoided if we have
the Maximum value of MaxDop set to 8 instead of 32 in this situation.

Could you please clarify on this one?

Thanks,
Dinesh M

**Comment by MikeRayMSFT** [comment]
Because the initial question was answered, I'm going to close this request. Feel free to open a new issue with the new question.

**Comment by pmasl** [comment]
FYI I added a lot of clarifying information to https://docs.microsoft.com/en-us/sql/relational-databases/thread-and-task-architecture-guide?view=sql-server-ver15#sql-server-task-scheduling and you can also check out the blog I wrote recently:  https://techcommunity.microsoft.com/t5/sql-server/what-is-maxdop-controlling/ba-p/1505968

**Comment by dinesh1440** [comment]
Hi

Thanks a lot for taking time to clarify. Really appreciate it. Thanks again.

Regards,
Dinesh M.
